### DISCUSSION TITLE
How to use "useConstantLod" for mesh

### DISCUSSION THREAD

**Discussion by PeterJBassett** [iTwin/community#23] [Q&A]
Hello I am developing the FutureOn-iTwin connector used by Shell.

There is a problem with tiling for a bathymetry mesh. The following is the guidance that was given from Bentley:

> When generating a .bim file, for Shell to use this new mode, they will need to create a RenderMaterial that uses a TextureMapping.Params with useConstantLod set to true. You can optionally specify its parameters in the ConstantLodProps.
> We previously communicated with Peter J. Bassett from FutureOn about this feature on GitHub discussions. We think he was involved in the FutureOn project of the .bim generator for Shell, but we are not 100% sure.

I have done the following in the connector code:

```
    const material = new RenderMaterialElement.Params(defaultMaterialName);
    material.diffuse = 0.9;
    material.patternMap = {
      TextureId: textureId,
      pattern_scale: [0.5, 0.5],
      pattern_useConstantLod: true,
    };
    const materialId = RenderMaterialElement.insert(iModelDb, definitionModelId, defaultMaterialName, material);
```
ie. I have added `pattern_useConstantLod: true,`

However, this has not fixed the problem with the mesh tiling.
Please can you let me know if I have misunderstood the guidance, and let me know how to fix the problem.

**Comment by pmconne** [comment]
Hi Peter,

What is the specific problem you are observing?
Do you know if the tiles are being produced by the mesh export service?
If they are, have you tried the default tiling?
Can you supply (via email to paul dot connelly at bentley dot com) the .bim file in question?

@MarcNeely can confirm whether your connector code is correct.

**Reply by PeterJBassett** [reply]
@pmconne here is a picture of the iModel that Shell are seeing, the tiling seems to be only partially drawn.
I don't have access to the bim file. I don't have access to Shell's iModels. I am sure that someone in Bentley will have access.

![imodel-mesh-problem](https://github.com/iTwin/community/assets/61423714/e9a416f5-bf81-4825-a56c-50ca879cc3b8)

**Reply by PeterJBassett** [reply]
@pmconne 
For your questions
> Do you know if the tiles are being produced by the mesh export service?

I don't know, and I don't know how to find out

> If they are, have you tried the default tiling?

I don't know how to do that either. I just tried to follow the instructions given by Bentley, as described above.

Also, I don't have access to the new mesh tiling feature in iModels/Plantsight. So I can't test the connector to see the problem.

**Comment by MarcNeely** [comment]
Hmm, I don't know.  Perhaps it's a spelling error?  It's pattern_usecontantlod (all lower-case), not pattern_useConstantLod.

**Reply by PeterJBassett** [reply]
Hello @MarcNeely I don't think that is right. 
According to the typescript it is `pattern_useConstantLod?: boolean;`

The typescript definition is as follows:
@itwin/core-common/MaterialProps
```
export interface TextureMapProps {
    /** Angle in degrees to rotate texture when applying; defaults to 0.0 if undefined */
    pattern_angle?: number;
    /** If true, flip the pattern map in U; if undefined, defaults to false */
    pattern_u_flip?: boolean;
    /** If true, flip the pattern map in V; if undefined, defaults to false */
    pattern_flip?: boolean;
    /** X, Y scale to apply to the pattern map; if undefined, defaults to {0,0}, which is almost never useful. */
    pattern_scale?: Point2dProps;
    /** X, Y offset to apply to the pattern map; if undefined, defaults to {0,0} */
    pattern_offset?: Point2dProps;
    /** Units to use when applying the scaling; if undefined, defaults to [[TextureMapUnits.Relative]] */
    pattern_scalemode?: TextureMapUnits;
    /** Mapping mode to use for the texture application; if undefined, defaults to [[TextureMapping.Mode.Parametric]] */
    pattern_mapping?: TextureMapping.Mode;
    /** Weight at which to combine diffuse image and color; if undefined, defaults to 1.0 */
    pattern_weight?: number;
    /** If true, override the mapping mode with constant LOD mapping for the normal map, defaults to false. */
    pattern_useConstantLod?: boolean;
    /** The number of times the texture is repeated if pattern_useConstantLod is true.  Increasing this will make the texture pattern appear smaller, decreasing it will make it larger. Defaults to 1.*/
    pattern_constantlod_repetitions?: number;
    /** An offset in world units used to shift the texture when pattern_useConstantLod is true. Defaults to (0, 0). */
    pattern_constantlod_offset?: Point2dProps;
    /** The minimum distance (from the eye to the surface) at which to clamp the texture size when pattern_useConstantLod is true. Defaults to 1. */
    pattern_constantlod_mindistanceclamp?: number;
    /** The maximum distance (from the eye to the surface) at which to clamp the texture size when pattern_useConstantLod is true. Defaults to 2^32. */
    pattern_constantlod_maxdistanceclamp?: number;
    /** The Id of the persistent [Texture]($backend) element defining the texture image. */
    TextureId: Id64String;
}
```

See https://github.com/iTwin/itwinjs-core/blob/9c6ca930ee6abfe860c16465a1bd0933448f00fe/core/common/src/MaterialProps.ts#L66

**Reply by pmconne** [reply]
@MarcNeely imodel-native and itwinjs-core [disagree](https://github.com/iTwin/imodel-native/blob/75d2c460a6a5d84b362e204f4aab6cd1bcf6c12a/iModelCore/iModelPlatform/PublicAPI/DgnPlatform/RenderMaterial.h#L68).

@PeterJBassett we will fix that. Meanwhile, can you try using the all-lowercase name and confirm that the constant LOD effect works?

**Reply by PeterJBassett** [reply]
Hello @pmconne I can try, but I expect that the typescript compiler will complain, I may be able to try setting the type to `any` to override the type checking.

**Reply by PeterJBassett** [reply]
Hello @pmconne Unfortunately I can't test the mesh tiles feature yet, becuase my Plantsight configuration still doesn't have this set up yet. 
I do have the "Generate Mesh" button, but I don't have the "Try It" / "Restore" button.

**Reply by pmconne** [reply]
No problem - thanks for trying. We can write a test that will replicate your problem and verify our fix. We'll let you know as soon as the fix becomes available. Sorry for the mix-up.

**Comment by pmconne** [comment]
@aruniverse fyi we know how to fix this, should not take long, but up to you whether or not to hold up the release for it.

**Reply by aruniverse** [reply]
we can try to sneak it into the release, but if not we can get it in a patch immediately after the release goes through
no rush on your end, take your time, but we'll try to get this out for users asap

**Comment by pmconne** [comment]
@PeterJBassett the fix has been incorporated into the 4.4.0 release candidate.

**Reply by PeterJBassett** [reply]
@pmconne ok thanks, I will update when it is available in npm
